Add workspace packages

JimMadge opened the pull request:
## :white_check_mark: Checklist

<!--
Thank you for your pull request!

- Before going any further, please review the [guidelines for contributing](../CONTRIBUTING.md) to this repository.
- If you're not yet ready to merge, please mark this pull request as a **draft** and add `'[WIP]'` to the title
- Replace the empty checkboxes [ ] below with checked ones [x] accordingly.
-->

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).

### :vertical_traffic_light: Depends on

<!--
List any other PRs that must be merged before this one
-->

- #1909 - Shouldn't dive too far into specifying packages/configuration until we have decided on the distro.
- #2034 

### :arrow_heading_up: Summary

<!--
Please explain what your pull request does here.
You might (optionally) want to include screenshots here.
-->

- Add ansible playbook for desired state
  - Install apt packages
  - Configures shell/prompts
- Remove FileUpload dynamic provider

I have moved some, but not all desired state here.
There were some files templated in cloud-init which need data from Pulumi/configuration.
We _can_ do that in Ansible, but we should think about how to we get the data to the workspaces.

- Template files locally and send them
- Pass variable to Ansible process?

There are some places where we could take advantage of the "official" Ansible community modules, or other packages from Ansible Galaxy. That will also require a bit of thought though as `ansible-galaxy` is really a package manager, so would be difficult to use in a locked down SRE.

#### Further Work

- Using Ansible galaxy collections/roles
- Testing (molecule?)
- Auditd configuration
- Clamav configuration
- Packages which require some manual work
  - RStudio
  - Nvidia drivers
  - Azure Data Studio and sqlcmd (possibly)
- Developer docs (how to change configuration and run using az CLI)
- Tightening networking rules
- Fix release lookup (and remove old releases)

### :closed_umbrella: Related issues

<!--
If your pull request will close any open issues (hopefully it will!) then add `Closes #<issue number>` here.
-->

Closes #1574

### :microscope: Tests

<!--
- Document any manual tests that you have carried out (*e.g.* deploying a new SHM and/or SRE) and confirm which commit you did this with
- Note that automated tests will be run as part of the CI process and will block this PR until they pass.
- Additionally, a successful code review may be required before this PR can be merged.
- If this pull request only changed documentation you can simply state 'Documentation only' here.
-->

Tested in a new deployment,

- Desired state container mounted correctly
- Playbook runs
- Systemd module installed
- Systemd module works
- Systemd timer installed and active
- Manually checked modified files, systemd services, _etc._
- Manually checked requested packages were installed

JimMadge committed:
WIP: Add workspace packages

JimMadge committed:
Add fd

JimMadge committed:
Enable bash completion

JimMadge committed:
Add default user bashrc

JimMadge committed:
Add desired state share

JimMadge commented:
Looking at some options to get the files on workspace VMs. The constraints are,

- Needs to happen early after deployment
  - Will have Linux utils and some basics from cloud-init (curl, bash, sed, but not az)
- Should happen on all workspaces
  - Easier to avoid authentication if possible
- Data is non-sensitive
  - No need for authentication for read access

The good solutions I can see are,

1. Blob container with anonymous read access
  - Use the existing data_configuration storage account (is that appropriate @jemrobinson)
  - Could create a new storage account if we are worried about enabling anonymous access
  - Anonymous read access
  - Networking rules only allow access from inside SRE
  - Upload a manifest file (as you need list permissions to list all files)
  - Script to fetch manifest, then fetch all files in the manifest using curl
2. Blob container with sftp access
  - Storage account needs [hierarchical namespace](https://learn.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-namespace)
  - Enable SFTP
  - Create local user with read/list permissions
    - With machine password for all workspaces or,
    - with ssh keypairs generated on each workspace
  - Script to scp from sftp using local account/password/pubkey

Non-options,

1. NFS: No easy way for us/admins to put files into NFS share
2. BlobFuse or SMB: more pre-requisites

Using anonymous access blob and http is nice for not needing authentication. However, you have to keep the manifest up to date which could be prone to error.
SFTP might be the best option, for the cost of storing the SFTP password and writing it to each workspace VM.

Thoughts @jemrobinson @craddm

jemrobinson commented:
Is the cloud-init Ansible support (https://cloudinit.readthedocs.io/en/latest/reference/modules.html#ansible) any help here?

Some thoughts on your suggestions:

1. `data_configuration` storage account
- Access is allowed from known `data_configuration_ip_addresses`
- Currently doesn't have any blob containers (only file shares)
- See e.g. the way we mount `ingress` and `egress` from blob storage over NFSv3

2. Enabling SFTP
- This is a bit of a mess, but @craddm got it working for a DSG in the past (hopefully notes on how to do it are somewhere)
- Not sure I would recommend it

How/when would the file get updated? Is it reasonable to imagine e.g. an Azure Function that would do the updating? If so, perhaps it could write directly to the VM and you wouldn't need a storage volume at all?

craddm commented:
I do have some notes on SFTP, yes. Wasn't enormously tricky, IIRC. Creating a local user was easy through the portal, and seems from Powershell it was easy to create one with a password rather than an SSH key.

I'm not seeing much about Azure Functions writing directly to VMs; you can trigger them on uploads to blob storage but so far I've only seen people suggest that the function copies files to a File Share

JimMadge commented:
@jemrobinson How much experience with key pairs in Azure/Pulumi do you have?

Looks like your can't generate a local user with password using Pulumi (at least, the password isn't an output).
You can specify an existing key pair though. That would mean we would need to be able to fetch the private key for each workspace.

jemrobinson commented:
What Pulumi resource are you using to generate a local user? I assume you don't mean in cloud-init?

JimMadge commented:
[azure_native.storage.LocalUser](https://www.pulumi.com/registry/packages/azure-native/api-docs/storage/localuser/)

Local user for SFTP, not for the workspaces.

JimMadge commented:
@jemrobinson Looks like Azure expects to keep the private key for keys in a keyvault.

I think the best option would be to use the [TLS package](https://www.pulumi.com/registry/packages/tls/) to generate a key, then store that as a secret.
(I'm now vaguely recalling using the TLS package before to do something like this).

jemrobinson commented:
I'm still a bit confused about what advantage we get for using SFTP over one of the other options. Is it just that we can do so earlier in the cloud-init order of operations? If so, why does this matter?

N.B. if you want to generate a Key in a Keyvault, I think you should probably use the [Key resource](https://www.pulumi.com/registry/packages/azure-native/api-docs/keyvault/key/)

JimMadge commented:
The SFTP way means,

- No requirements beyond what is already on a bare Ubuntu VM image
- No need to have a manifest (you can just `scp -r user@host:* /`)

The current downside is it means sharing a private key between workspaces (or creating a new local account with key pair for each workspace).

JimMadge committed:
Merge remote-tracking branch 'origin/develop' into workspace_software

JimMadge committed:
WIP add desired state container

JimMadge committed:
WIP: Add script to fetch and run ansible

JimMadge committed:
WIP: Add desired state systemd timer

JimMadge committed:
Fix typing errors

JimMadge committed:
Upload ansible to desired state container

JimMadge committed:
Move workspace packages from cloudinit to ansible

JimMadge committed:
Write nslcd.conf with ansible

JimMadge committed:
Write xession skel file with ansible

JimMadge committed:
Add missing files

JimMadge committed:
Edit nsswitch.conf in ansible

JimMadge committed:
Correct regex

JimMadge committed:
Move PAM configuration to ansible

jemrobinson commented on data_safe_haven/resources/workspace/workspace.cloud_init.mustache.yaml:
```diff
@@ -48,10 +20,44 @@ write_files:
+  - path: "/etc/systemd/system/desired-state.timer"
+    permissions: "0644"
+    content: |
+      [Unit]
+      Description=Run desired state script daily
+
+      [Timer]
+      OnCalendar=daily
+      Persistent=true
```
Is daily often enough? What time will this happen? It might be a while after first deployment.

JimMadge replied:
Feels OK as a baseline. In the long term, I think we can trigger this manually.

I believe daily is midnight, plus whatever default random delay (we can specify that if we want to avoid running too much at once).

I was going to say the service is run at the end of cloud init. Then I realised I had forgotten to do that, so I've added it now 😅.

JimMadge committed:
Revert "Write nslcd.conf with ansible"

This reverts commit cb8f1a87c905eae8f4c37ffd685f80027358c510.

Return nslcd configuration to cloudinit as it depends on parameters
from pulumi.

JimMadge committed:
Move sshd configuration to ansible

JimMadge committed:
Remove (probably) unneeded systemctl call

JimMadge committed:
Enable and start xrdp services

JimMadge committed:
Merge remote-tracking branch 'origin/develop' into workspace_software

JimMadge committed:
Merge remote-tracking branch 'origin/develop' into workspace_software

JimMadge committed:
Merge branch 'fix_firewall_priority' into workspace_software

JimMadge committed:
Correct desired state container args

JimMadge committed:
Use valid container name

JimMadge committed:
Add resource name

JimMadge committed:
Add service to permission scope args

JimMadge committed:
Use apply/lambda for b64encoded key

JimMadge committed:
Use valid username

JimMadge committed:
Use identity in scp command

JimMadge committed:
Don't b64encode input

JimMadge committed:
Pass Pulumi outputs to cloud init

JimMadge committed:
Remove boilerplate in cloudint templating

JimMadge committed:
Move test file provisioning to Ansible

JimMadge committed:
Remove Windows VM Props

JimMadge committed:
Improve and test cloudinit templating functions

JimMadge committed:
WIP use NFSv3 blob for desired state

JimMadge committed:
Correct account name

JimMadge committed:
Correct allowed IP addresses and subnet

JimMadge committed:
Allow access to config subnet from workspaces

JimMadge committed:
Correct priority

JimMadge committed:
Upload desired state files rather than archive

JimMadge committed:
Use str for _name

JimMadge committed:
Correct storage account name

JimMadge committed:
Give uploaded blobs unique names

JimMadge committed:
Correct blob_name argument

JimMadge committed:
Add BlobAccessTier

JimMadge committed:
Use relative paths for blob names

JimMadge committed:
Use apply for promises

jemrobinson commented on data_safe_haven/infrastructure/programs/sre/data.py:
```diff
@@ -550,10 +550,14 @@ def __init__(
-            file_name = file_path.name
-            blob_name = file_path.relative_to(
-                (resources_path / "workspace" / "ansible").absolute()
+            file_path = Output.from_input(file).apply(lambda file: Path(file.path))
```
```suggestion
            file_path = Output.from_input(file).apply(Path)
```
might work here?

JimMadge committed:
Avoid unnecessary promises

JimMadge committed:
Remove access tier

Tiers don't appear to be valid for this container/account type. Despite
being optional in the docs, an argument seems to be expected. Perhaps a
bug fixed in a newer version of pulumi-azure-native.

JimMadge commented:
## Problem

Blob storage accounts with NFSv3 enabled don't (seem to) use access tiers for blobs.
When uploading blobs

https://github.com/alan-turing-institute/data-safe-haven/blob/9ac1ee022f94a8ce6a147785f234f6321cb1e532/data_safe_haven/infrastructure/programs/sre/data.py#L554-L562

If [`access_tier`](https://www.pulumi.com/registry/packages/azure-native/api-docs/storage/blob/#inputs) is undefined, it defaults to `None`.
This causes an error `ValueError: '' is not a valid BlobAccessTier`.
Settings `access_tier` to any `storage.BlobAccessTier` enum value, or string causes an error

```
creating replacement (0s) ; error: updating access tier for blob
"hello.txt" (container "desiredstate" / account
"shdaisrehojdesiredstatec"): blobs.Client#SetTier: Failure responding to
request: StatusCode=400 -- Original Error: autorest/azure: Service
returned an error. Status=400 Code="InvalidQueryParameterValue"
Message="Value for one of the query parameters specified in the request
URI is
invalid.
```

Perhaps the Pulumi azure provider is always calling the Azure API method for setting access tier.
For this storage account/container, however, there is no valid value for access tier.

### Thoughts

This may be an azure-pulumi-native bug, or at least an edge case that isn't supported.
The Azure API, through az CLI, can handle uploading blobs to this kind of container. _I.e._ the setting of access tier isn't an inherent part of uploading a blob. We could create a custom Pulumi resource which uses the azure Python SDK to upload a blob.

JimMadge commented:
Importing a blob uploaded in the portal gives

```Python
requirementblob = azure_native.storage.Blob("requirementblob",
    access_tier=azure_native.storage.BlobAccessTier(""),
    account_name="shdaisrehojdesiredstatec",
    blob_name="requirements.txt",
    container_name="desiredstate",
    content_md5="b4s63roBt3c9gnki/yIG8A==",
    content_type="text/plain",
    metadata={},
    resource_group_name="shm-daimyo-sre-hojo-rg-data",
    type=azure_native.storage.BlobType.BLOCK,
    opts=pulumi.ResourceOptions(protect=True))
```

Using `access_tier=azure_native.storage.BlobAccessTier("")` gives the same error as when using `None` or providing no argument, `ValueError: '' is not a valid BlobAccessTier`.

JimMadge committed:
Merge remote-tracking branch 'origin/develop' into workspace_software

JimMadge committed:
Update Pulumi Azure Native minimum version

This release contains the fix for blobs in NFSv3 enabled storage
accounts.

JimMadge committed:
Merge remote-tracking branch 'origin/develop' into workspace_software

JimMadge committed:
Correct desired state NFS address

JimMadge committed:
Add NSG rule for desired state

Outbound connection allowed from workspaces subnet to configuration data
subnet

JimMadge committed:
Lint ansible

JimMadge committed:
Add ansible lint script

JimMadge committed:
Rename systemd_service module

The version on ansible on the workspace VMs uses the old module name
"systemd".

JimMadge committed:
Correct handler name

JimMadge committed:
Correct sshd config validate command

JimMadge committed:
Update script to run ansible from correct dir

JimMadge committed:
Correct path

JimMadge committed:
Move tests to dshadmin home dir

JimMadge committed:
Remove unused function

JimMadge committed:
Update test

jemrobinson commented on data_safe_haven/infrastructure/programs/sre/data.py:
```diff
@@ -456,6 +457,154 @@ def __init__(
+        BlobContainerAcl(
+            f"{container_desired_state._name}_acl",
+            BlobContainerAclProps(
+                acl_user="rwx",
+                acl_group="r-x",
+                acl_other="r-x",
```
Why do we want these permissions?

JimMadge replied:
We probably actually want something like 550.

rx is needed for root to read and browse. Others shouldn't need it though.

jemrobinson commented on data_safe_haven/infrastructure/programs/sre/networking.py:
```diff
@@ -355,6 +355,18 @@ def __init__(
                     source_address_prefix=SREIpRanges.user_services_software_repositories.prefix,
                     source_port_range="*",
                 ),
+                network.SecurityRuleArgs(
+                    access=network.SecurityRuleAccess.ALLOW,
+                    description="Allow inbound connections from workspaces.",
+                    destination_address_prefix=SREIpRanges.data_configuration.prefix,
+                    destination_port_range="*",
+                    direction=network.SecurityRuleDirection.INBOUND,
+                    name="AllowWorkspacesInbound",
+                    priority=NetworkingPriorities.INTERNAL_SRE_WORKSPACES,
+                    protocol=network.SecurityRuleProtocol.ASTERISK,
+                    source_address_prefix=SREIpRanges.workspaces.prefix,
+                    source_port_range="*",
+                ),
```
I think we don't want this. The configuration data is deliberately kept isolated from user space so that we can store secrets etc. there as required.

Since you've set up a new storage account container for desired state, you can attach the private endpoint to its own subnet.

jemrobinson commented on data_safe_haven/infrastructure/programs/sre/networking.py:
```diff
@@ -1282,6 +1294,18 @@ def __init__(
+                    protocol=network.SecurityRuleProtocol.ASTERISK,
+                    source_address_prefix=SREIpRanges.workspaces.prefix,
+                    source_port_range="*",
+                ),
```
As above, this should be a link to a dedicated desired state endpoint.

jemrobinson commented on data_safe_haven/infrastructure/programs/sre/workspaces.py:
```diff
@@ -215,34 +186,13 @@ def __init__(
-            cloudinit = chevron.render(f_cloudinit, mustache_values)
-            return b64encode(cloudinit)
+            cloudinit = chevron.render(f_cloudinit, kwargs)
+            logger.debug(f"Generated cloud-init config:\n {cloudinit}")
```
Are we worried about exposing secrets in the logs here?

JimMadge replied:
I'm not particularly worried. There aren't any secrets there (at the moment). It would also only expose these to a user who can read all of the secrets already.

jemrobinson replied:
OK, and do we want a multiline log message here? I thought we were trying to avoid those.

jemrobinson commented on data_safe_haven/infrastructure/programs/sre/workspaces.py:
```diff
@@ -131,9 +132,13 @@ def __init__(
+        ).apply(lambda kwargs: self.template_cloudinit(**kwargs))
+        b64cloudinit = Output.from_input(cloudinit).apply(
+            lambda cloudinit: b64encode(cloudinit)
+        )
```
You shouldn't need `Output.from_input` as cloudinit should be an `Output[str]` already.

```suggestion
        b64cloudinit = cloudinit.apply(b64encode)
```

jemrobinson commented on data_safe_haven/resources/workspace/ansible/host_vars/localhost.yaml:
```diff
@@ -0,0 +1,122 @@
+    jammy:
+    noble:
+  - category: development
+    common:
```
Missing wrt 4.X.Y
- java
- rustc
- scala
- spark-shell

JimMadge replied:
- Do we want to keep Java, scala? I would assume they are much less common
- rustc will be installed as part of cargo (we could add it explicitly though)
- Is spark-shell any use given we don't have a spark server?

jemrobinson replied:
I think supporting as many programming languages as possible is part of our aim. Java and scala should be in Ubuntu's repos, so I imagine that's quite easy to support (`maven` has been a bit of a problem with scala projects in the past, but we can cross that bridge if we come to it again). Agreed that rustc should be installed by cargo, but let's be explicit. We can drop spark-shell if it's useless without a cluster (I don't know whether it can be run standalone).

Are there any other languages we should be considering here? e.g. dotnet/mono support?

JimMadge replied:
Agreed. I think in practice Python > R >> everything else, but it shouldn't be too difficult to support more.

Didn't we find dotnet had some daemon that wanted to phone home?

JimMadge replied:
We are missing functional languages, LISP dialects, Haskell, F#.

jemrobinson replied:
F# would be covered by dotnet (I think - check with e.g. Evelina). If you can find a LISP/Haskell practitioner and find out what they need installed then yes, let's do that.

jemrobinson replied:
> Didn't we find dotnet had some daemon that wanted to phone home?

I think you can (a) turn off dotnet telemetry and (b) ignore it, since failure to phone home doesn't break anything.

Maven was more of a problem, since scala programs wouldn't run until they'd updated all their packages.

jemrobinson commented on data_safe_haven/resources/workspace/ansible/host_vars/localhost.yaml:
```diff
@@ -0,0 +1,122 @@
+      - python3.12
+      - python3.12-venv
+      - python-is-python3
+  - category: editors
```
Missing wrt 4.X.Y
- code (this is what pretty much all our users use!)
- nano
- pycharm-community
- rstudio

JimMadge replied:
- Added codium and pycharm-community as snaps.
- Nano is part of the base image but I've added explicitly

Rstudio is more difficult as posit host debs on their own site and don't distribute on the canonical repos or as a snap.
Can we make move that to a new issue?

jemrobinson replied:
Yes, any/all of these can be put into new issues, together or separately as you prefer. This was just to ensure that anything we're removing is done deliberately.

jemrobinson commented on data_safe_haven/resources/workspace/ansible/host_vars/localhost.yaml:
```diff
@@ -0,0 +1,122 @@
+    focal:
+    jammy:
+    noble:
+  - category: office
```
Missing wrt 4.X.Y
- latex
- xelatex

JimMadge replied:
Added texlive-full.

jemrobinson commented on data_safe_haven/resources/workspace/ansible/host_vars/localhost.yaml:
```diff
@@ -0,0 +1,122 @@
+    jammy:
+    noble:
+  - category: database
+    common:
+      - libpq-dev     # interact with PostgreSQL databases
+      - msodbcsql17   # interact with Microsoft SQL databases
+      - unixodbc-dev  # interact with Microsoft SQL databases
```
Missing wrt 4.X.Y
- azuredatastudio
- dbeaver-ce
- psql
- sqlcmd

JimMadge replied:
Added dbeaver-ce as a snap and postgreql-client-common (which provides the psql command).

Azure Data Studio and sqlcmd are more difficult, requiring custom PPAs or fetching debs.
We could open a new issue, I feel like they overlap with existing tools though. I would lean towards not including them.

jemrobinson commented on data_safe_haven/resources/workspace/ansible/host_vars/localhost.yaml:
```diff
@@ -0,0 +1,122 @@
+    focal:
+    jammy:
+    noble:
+  - category: utilities
+    common:
+      - bat
+      - fd-find
+      - fzf
+      - htop
+      - httpie
+      - nvtop
+      - ripgrep
+      - rsync
+      - tmux
+      - wget
```
Missing wrt 4.X.Y
- git
- nvidia-smi
- weka

jemrobinson replied:
re: `weka` there was an issue [a couple of years ago](https://github.com/alan-turing-institute/data-safe-haven/issues/710) with the version available through Ubuntu's repos being quite old. Worth checking whether this is still an issue.

JimMadge replied:
Git was already there, I've added `git-all` though to include other modules (lfs and so on). I've added weka.

nvidia-smi should probably be installed as part of `nvidia-utils-???`. You need to specify a version, which really requires knowing which driver version your hardware needs. I think we should open an issue for that.

jemrobinson committed:
Merge branch 'develop' into workspace_software

JimMadge committed:
Update data_safe_haven/infrastructure/programs/sre/workspaces.py

Co-authored-by: James Robinson <james.em.robinson@gmail.com>

JimMadge committed:
Add rustc package explicitly

JimMadge committed:
Add more editors/IDEs

JimMadge committed:
Add full Tex Live distribution

jemrobinson commented:
Deployments from this branch are failing with the following error:

```
 +  azure-native:network:VirtualNetwork sre_networking_virtual_network **creating failed** error: Code="InvalidResourceReference" Message="Resource
/subscriptions/3f1a8e26-eae2-4539-952a-0a6184ec248a/resourceGroups/SHM-GREEN-SRE-FERN-RG-NETWORKING/providers/Microsoft.Network/routeTables/SHM-GREEN-SRE-FERN-ROUTE-TABLE referenced by resource
/subscriptions/3f1a8e26-eae2-4539-952a-0a6184ec248a/resourceGroups/shm-green-sre-fern-rg-networking/providers/Microsoft.Network/virtualNetworks/shm-green-sre-fern-vnet was not found. Please make sure that the referenced resource exists, and that both resources are in
the same region." Details=[{"code":"NotFound","message":"Resource /subscriptions/3f1a8e26-eae2-4539-952a-0a6184ec248a/resourceGroups/SHM-GREEN-SRE-FERN-RG-NETWORKING/providers/Microsoft.Network/routeTables/SHM-GREEN-SRE-FERN-ROUTE-TABLE not found."}]
```

@craddm @JimMadge have you seen this?

JimMadge committed:
Add more database tools

JimMadge committed:
Add weka and more git tools

JimMadge commented:
@jemrobinson I've not seen that :eyes:.

JimMadge committed:
Add packages required for ansible linting

JimMadge committed:
Remove multiline log message

JimMadge committed:
Add java and scala

JimMadge committed:
Add dotnet and mono support

jemrobinson commented on data_safe_haven/resources/workspace/ansible/host_vars/localhost.yaml:
```diff
@@ -31,15 +31,18 @@ package_categories:
       - gfortran
       - git-all
       - golang
+      - mono-devel
```
Do we want dotnet AND mono? It's been a while since I had to think about this, but aren't they doing the same thing?

JimMadge replied:
I'm not familiar with them. We could be opinionated at say mono only, but both are available so it seems like a low cost. I'm not sure how compatible they are.

This is a bit different to Matlab/Octave where there would also be licensing problems.

jemrobinson replied:
IIRC Mono was an open-source implementation of DotNet when it didn't support Linux. Subsequently, Mono was bought by Microsoft and DotNet core is available for Linux. I think we probably don't need Mono here - can you check if that's right?

JimMadge replied:
> IIRC Mono was an open-source implementation of DotNet when it didn't support Linux.

That might be true but I think there are still an important difference due to licenses and that .Net core doesn't implement all .Net functions (_e.g._ silverlight/moonlight).

I would suspect they are very similar but there would be some edge cases which might require one implementation in particular.
If we were to remove one, I would lean towards keeping on Mono as it is FOSS.

jemrobinson replied:
Wasn't Silverlight deprecated in the 2010s?

No strong preference between the two - can you ask someone who actually uses a .NET language which is better?

JimMadge replied:
Can you point me in the direction of someone who uses .NET :laughing:? (I'm honestly not aware of anyone).

jemrobinson replied:
@evelinag?

JimMadge replied:
Looks like .NET is the way to go. It is MIT now and more actively developed.

We can worry about edge cases where Mono is needed over .NET when they happen :+1:

JimMadge committed:
Add more development tools/languages

JimMadge committed:
Add comments to development packages

jemrobinson commented:
@JimMadge I'm getting the `python inline source runtime error: '' is not a valid BlobAccessTier` message. Do we need an updated version of `pulumi` or `pulumi-azure-native` to get the fix? If so, can you add the required change to this PR?

JimMadge commented:
@jemrobinson Which versions do you have?

I think `pulumi-azure-native` should be up to date in the requirements, and I _think_ that is all we need.

I also found that if the stack already had a blob tier defined I would still get the error. Removing that from the stack or deleting and recreating the resources fixed it.

JimMadge committed:
Improve desired state ACLs

JimMadge committed:
Add desired state data subnet

JimMadge committed:
Remove Mono

JimMadge committed:
Fix SRE IP ranges test

JimMadge committed:
Fix guacamole containers ip test

JimMadge committed:
Fix application gateway backend test

JimMadge committed:
Run lint:fmt

jemrobinson commented on data_safe_haven/infrastructure/programs/sre/networking.py:
```diff
@@ -1282,6 +1341,18 @@ def __init__(
                     source_address_prefix=SREIpRanges.workspaces.prefix,
                     source_port_range="*",
                 ),
+                network.SecurityRuleArgs(
+                    access=network.SecurityRuleAccess.ALLOW,
+                    description="Allow outbound connections to configuration data endpoints.",
+                    destination_address_prefix=SREIpRanges.data_configuration.prefix,
+                    destination_port_range="*",
+                    direction=network.SecurityRuleDirection.OUTBOUND,
+                    name="AllowDataConfigurationEndpointsOutbound",
+                    priority=NetworkingPriorities.INTERNAL_SRE_DATA_CONFIGURATION,
+                    protocol=network.SecurityRuleProtocol.ASTERISK,
+                    source_address_prefix=SREIpRanges.workspaces.prefix,
+                    source_port_range="*",
+                ),
```
```suggestion
                network.SecurityRuleArgs(
                    access=network.SecurityRuleAccess.ALLOW,
                    description="Allow outbound connections to desired state data endpoints.",
                    destination_address_prefix=SREIpRanges.data_desired_state.prefix,
                    destination_port_range="*",
                    direction=network.SecurityRuleDirection.OUTBOUND,
                    name="AllowDataDesiredStateEndpointsOutbound",
                    priority=NetworkingPriorities.INTERNAL_SRE_DATA_DESIRED_STATE,
                    protocol=network.SecurityRuleProtocol.ASTERISK,
                    source_address_prefix=SREIpRanges.workspaces.prefix,
                    source_port_range="*",
                ),
```

N.B. This will also need a new entry in the `NetworkingPriorities` enum.

By the way, do we not know which destination port this connection will use? If we do, we should add that to this rule.

JimMadge replied:
I knew I would miss something :smile:, quite a few places to make changes.

JimMadge replied:
> The NFS 3.0 protocol uses ports 111 and 2049. If you're connecting from an on-premises network, make sure that your client allows outgoing communication through those ports. If you have granted access to specific VNets, make sure that any network security groups associated with those VNets don't contain security rules that block incoming communication through those ports.

https://learn.microsoft.com/en-us/azure/storage/blobs/network-file-system-protocol-support#supported-network-connections

JimMadge replied:
Fixed in ed11e31 and 69674c2

JimMadge committed:
Update data_safe_haven/infrastructure/programs/sre/networking.py

Co-authored-by: James Robinson <james.em.robinson@gmail.com>

JimMadge committed:
Add desired state network priority

JimMadge committed:
Add port definitions for Azure Blob NFSv3 access

JimMadge committed:
Correct package name

jemrobinson commented on data_safe_haven/infrastructure/programs/sre/networking.py:
```diff
@@ -1274,14 +1333,26 @@ def __init__(
                     description="Allow outbound connections to private data endpoints.",
                     destination_address_prefix=SREIpRanges.data_private.prefix,
-                    destination_port_range="*",
+                    destination_port_range=[Ports.NFS_V3_1, Ports.NFS_V3_2],
```
```suggestion
                    destination_port_ranges=[Ports.NFS_V3_1, Ports.NFS_V3_2],
```

jemrobinson committed:
:bug: Fix port ranges specifiers

jemrobinson committed:
:bug: Fix port ranges specifiers

jemrobinson commented on data_safe_haven/infrastructure/programs/sre/networking.py:
```diff
@@ -396,6 +396,65 @@ def __init__(
+                    access=network.SecurityRuleAccess.ALLOW,
+                    description="Allow inbound connections from workspaces.",
+                    destination_address_prefix=SREIpRanges.data_configuration.prefix,
+                    destination_port_range=[Ports.NFS_V3_1, Ports.NFS_V3_2],
```
```suggestion
                    destination_port_ranges=[Ports.NFS_V3_1, Ports.NFS_V3_2],
```

jemrobinson commented on data_safe_haven/infrastructure/programs/sre/networking.py:
```diff
@@ -406,7 +465,7 @@ def __init__(
                     description="Allow inbound connections from SRE workspaces.",
                     destination_address_prefix=SREIpRanges.data_private.prefix,
-                    destination_port_range="*",
+                    destination_port_range=[Ports.NFS_V3_1, Ports.NFS_V3_2],
```
```suggestion
                    destination_port_ranges=[Ports.NFS_V3_1, Ports.NFS_V3_2],
```

jemrobinson commented:
Currently hitting a problem with an incorrect type being passed to Pulumi:

```
  pulumi:pulumi:Stack (data-safe-haven-shm-green-sre-fern):
    error: python inline source runtime error: Unexpected type. Expected 'list' got '<class 'str'>'
    Traceback (most recent call last):
      File "/Users/jrobinson/Library/Application Support/hatch/env/virtual/data-safe-haven/Svgq2p76/data-safe-haven/lib/python3.12/site-packages/pulumi/automation/_server.py", line 80, in run
        loop.run_until_complete(run_in_stack(self.program))
      File "/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/asyncio/base_events.py", line 687, in run_until_complete
        return future.result()
               ^^^^^^^^^^^^^^^
      File "/Users/jrobinson/Library/Application Support/hatch/env/virtual/data-safe-haven/Svgq2p76/data-safe-haven/lib/python3.12/site-packages/pulumi/runtime/stack.py", line 138, in run_in_stack
        await run_pulumi_func(run)
      File "/Users/jrobinson/Library/Application Support/hatch/env/virtual/data-safe-haven/Svgq2p76/data-safe-haven/lib/python3.12/site-packages/pulumi/runtime/stack.py", line 52, in run_pulumi_func
        await wait_for_rpcs()
      File "/Users/jrobinson/Library/Application Support/hatch/env/virtual/data-safe-haven/Svgq2p76/data-safe-haven/lib/python3.12/site-packages/pulumi/runtime/stack.py", line 114, in wait_for_rpcs
        await task
      File "/Users/jrobinson/Library/Application Support/hatch/env/virtual/data-safe-haven/Svgq2p76/data-safe-haven/lib/python3.12/site-packages/pulumi/runtime/resource.py", line 909, in do_register
        resolver = await prepare_resource(res, ty, custom, remote, props, opts, typ)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/Users/jrobinson/Library/Application Support/hatch/env/virtual/data-safe-haven/Svgq2p76/data-safe-haven/lib/python3.12/site-packages/pulumi/runtime/resource.py", line 196, in prepare_resource
        serialized_props = await rpc.serialize_properties(
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/Users/jrobinson/Library/Application Support/hatch/env/virtual/data-safe-haven/Svgq2p76/data-safe-haven/lib/python3.12/site-packages/pulumi/runtime/rpc.py", line 215, in serialize_properties
        result = await serialize_property(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/Users/jrobinson/Library/Application Support/hatch/env/virtual/data-safe-haven/Svgq2p76/data-safe-haven/lib/python3.12/site-packages/pulumi/runtime/rpc.py", line 359, in serialize_property
        await serialize_property(
      File "/Users/jrobinson/Library/Application Support/hatch/env/virtual/data-safe-haven/Svgq2p76/data-safe-haven/lib/python3.12/site-packages/pulumi/runtime/rpc.py", line 521, in serialize_property
        k: await serialize_property(
           ^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/Users/jrobinson/Library/Application Support/hatch/env/virtual/data-safe-haven/Svgq2p76/data-safe-haven/lib/python3.12/site-packages/pulumi/runtime/rpc.py", line 355, in serialize_property
        element_type = _get_list_element_type(typ)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/Users/jrobinson/Library/Application Support/hatch/env/virtual/data-safe-haven/Svgq2p76/data-safe-haven/lib/python3.12/site-packages/pulumi/runtime/rpc.py", line 168, in _get_list_element_type
        raise AssertionError(f"Unexpected type. Expected 'list' got '{typ}'")
    AssertionError: Unexpected type. Expected 'list' got '<class 'str'>'
    error: Exception calling application: invalid state
```

can you see where this is happening @JimMadge ?

jemrobinson committed:
Merge branch 'develop' into workspace_software

jemrobinson committed:
Merge branch 'develop' into workspace_software

JimMadge commented:
@jemrobinson I didn't see that before, so it is likely it was introduce during the networking changes,

- Adding desired state subnet
- Changing ports in NSG rules

Is there no more output? When I had similar errors I think there was often a clue to which line in our code had a problem.

JimMadge committed:
Merge remote-tracking branch 'refs/remotes/origin/workspace_software' into workspace_software

jemrobinson commented:
I think I've fixed it with 39aa441.

JimMadge commented:
Sorry, I didn't catch the argument for a port string or list of strings was different 😅.

jemrobinson commented on data_safe_haven/infrastructure/programs/sre/networking.py:
```diff
@@ -399,6 +399,65 @@ def __init__(
         )
+        nsg_data_desired_state = network.NetworkSecurityGroup(
+            f"{self._name}_nsg_data_desired_state",
+            network_security_group_name=f"{stack_name}-nsg-data-desired-state",
```
```suggestion
            location=props.location,
            network_security_group_name=f"{stack_name}-nsg-data-desired-state",
```

jemrobinson committed:
:bug: Add missing 'location' property to NSG

jemrobinson committed:
:art: Inline b64cloudinit which is only used in one place

jemrobinson commented:
OK, I've now got this deployed, but I can't connect to the Workspace from the Guacamole dashboard. Is this working for you @JimMadge ?

jemrobinson commented on data_safe_haven/infrastructure/programs/sre/data.py:
```diff
@@ -456,6 +461,154 @@ def __init__(
+        storage_account_data_desired_state_endpoint = network.PrivateEndpoint(
+            f"{storage_account_data_desired_state._name}_private_endpoint",
+            location=props.location,
+            private_endpoint_name=f"{stack_name}-pep-storage-account-desired-state",
```
```suggestion
            private_endpoint_name=f"{stack_name}-pep-storage-account-data-desired-state",
```

jemrobinson commented:
N.B. This is caused by a failure to mount the `/desiredstate` volume. Could be due to overly-restrictive NSG rules (e.g. the recent change to which ports are allowed)? Can you take a look @JimMadge ?

```
root@shm-green-sre-fern-vm-workspace-01:~# mount -fav
/                        : ignored
/boot/efi                : already mounted
mount.nfs: timeout set for Thu Jul 18 13:30:34 2024
mount.nfs: trying text-based options 'sec=sys,vers=3,nolock,proto=tcp,addr=10.0.1.20'
mount.nfs: prog 100003, trying vers=3, prot=6
mount.nfs: trying 10.0.1.20 prog 100003 vers 3 prot TCP port 2048
```

```
root@shm-green-sre-fern-vm-workspace-01:~# nslookup shgresreferdesiredstatec.blob.core.windows.net
Server:         127.0.0.53
Address:        127.0.0.53#53

Non-authoritative answer:
shgresreferdesiredstatec.blob.core.windows.net  canonical name = shgresreferdesiredstatec.privatelink.blob.core.windows.net.
Name:   shgresreferdesiredstatec.privatelink.blob.core.windows.net
Address: 10.0.1.20
```

JimMadge committed:
Update data_safe_haven/infrastructure/programs/sre/data.py

Co-authored-by: James Robinson <james.em.robinson@gmail.com>

jemrobinson commented on data_safe_haven/resources/workspace/workspace.cloud_init.mustache.yaml:
```diff
@@ -75,57 +107,17 @@ apt:
-  - systemctl start xrdp
-  - systemctl start xrdp-sesman
 
   # Mount all external volumes
```
Add a test for /desiredstate here too.

N.B. Is there a better place we can put this than as a new root-level directory? There must be an established location for config files that would be a better mount point.

JimMadge replied:
I would probably put all of the mounts in `/mnt` when is conventionally used for non-system disks at network drives. (`/media` or more recently `/run/media` for removable media).

Do you want to open an issue?

jemrobinson replied:
Captured in #2027 and marked for rc3.

jemrobinson commented:
Looks like you may have added the wrong NFS ports. [This doc](https://learn.microsoft.com/en-us/azure/storage/blobs/network-file-system-protocol-support-how-to) says we need 111 and 2048.

JimMadge commented:
@jemrobinson 🤦 I think we are trying to debug at the same time. Saw the port change from 2049 to 2048 before my eyes 😆.

JimMadge committed:
Correct NFS port

JimMadge committed:
Add check for desired state mount

jemrobinson commented on data_safe_haven/resources/workspace/workspace.cloud_init.mustache.yaml:
```diff
@@ -75,65 +107,34 @@ apt:
   - while (! mountpoint -q /home); do sleep 5; mount /home; done
   - while (! mountpoint -q /output); do sleep 5; mount /output; done
   - while (! mountpoint -q /shared); do sleep 5; mount /shared; done
-  - findmnt
```
What's the reason for dropping `findmnt`? Is it because we think sysadmins will need/want to look at these log files? Or is this done better somewhere else?

JimMadge replied:
An accident 😅.

jemrobinson commented:
The mount is still failing. I'm going to try with `other: r-x` permissions as I'm not sure which user the storage account is mounted with.

jemrobinson committed:
:bug: Fix storage container ACLs

jemrobinson committed:
:bug: Ignore custom_dns_configs changes or PrivateEndpoint will be redeployed on each deploy() call

jemrobinson commented:
Container mounting is fixed but GUI login is still not working. Running commands on the serial console indicates ansible probably hasn't run:

```
dshadmin@shm-green-sre-fern-vm-workspace-01:~$ which rustc
dshadmin@shm-green-sre-fern-vm-workspace-01:~$ which cargo
dshadmin@shm-green-sre-fern-vm-workspace-01:~$ 
dshadmin@shm-green-sre-fern-vm-workspace-01:~$ systemctl status desired-state
○ desired-state.service - Run desired state script
     Loaded: loaded (/etc/systemd/system/desired-state.service; static)
     Active: inactive (dead)
```

any thoughts on how to debug this @JimMadge ?

JimMadge committed:
Restore findmnt

JimMadge commented:
> Container mounting is fixed but GUI login is still not working. Running commands on the serial console indicates ansible probably hasn't run:
> 
> ```
> dshadmin@shm-green-sre-fern-vm-workspace-01:~$ which rustc
> dshadmin@shm-green-sre-fern-vm-workspace-01:~$ which cargo
> dshadmin@shm-green-sre-fern-vm-workspace-01:~$ 
> dshadmin@shm-green-sre-fern-vm-workspace-01:~$ systemctl status desired-state
> ○ desired-state.service - Run desired state script
>      Loaded: loaded (/etc/systemd/system/desired-state.service; static)
>      Active: inactive (dead)
> ```
> 
> any thoughts on how to debug this @JimMadge ?

The `systemctl` output is what I would expect. It is a one-shot service so one the command has finished the unit won't be active.

Looking at `journalctl -u desired-state.service` should show the output. You can run the script at `/root/desired_state.sh`.
It's also quite easy to make changes because you can use `az` to update the ansible project on a deployed environment.

```command
az storage blob upload-batch --account-name <storageaccount> -d <container> -s ./ --overwrite
```

JimMadge committed:
Run lint:fmt

JimMadge committed:
Correct popd command

JimMadge commented:
@jemrobinson It is working now (and probably _was_ before although there was a non-zero exit code).

The problem fixed in 1713ae5 fixed the exit code.

Was the output you posted above from a VM that had been deployed while the mount was broken?
I logged into a deployed VM and rustc and cargo were installed. Running Ansible made no changes.
I suspect at deployment the NFS share wasn't mounted, so the initial run of `desired-state.service` didn't work. But the timer ran it overnight, so when I looked this morning it was all configured.

jemrobinson commented:
@JimMadge: I think we need to EITHER wait until the mount is definitely available before running the one-shot service (I *thought* this should be what the existing cloud-init code is doing but it doesn't seem to be) OR add an explicit dependency in the service that the mount exists (I haven't tried this, but I think it's something like `After=mnt-whatever.mount`)

JimMadge commented:
Yes it should do.

Looking at a fresh deployment, the timer isn't active although `systemctl enable desired-state.service` should have run. It might be that `systemctl daemon-reload` wasn't run after writing unit files.

JimMadge committed:
Reload systemd daemon in cloud init

JimMadge committed:
Send desired state output to term and journal

JimMadge commented:
@jemrobinson it looks like cloud-init isn't getting past the mount stage.

```
>=== Mounting all external volumes... ===<
  /etc/fstab  # CLOUD_IMG: This file was created/modified by the Cloud Image build process
  /etc/fstab  UUID=fd0e4ffa-6c5f-49cc-9a56-46113512bb1a /        ext4   discard,errors=remount-ro       0 1
  /etc/fstab  UUID=6C4C-3E9C    /boot/efi       vfat    umask=0077      0 1
  /etc/fstab  shdaisreshidesiredstatec.blob.core.windows.net:/shdaisreshidesiredstatec/desiredstate     /desired_state  nfs     ro,_netdev,sec=sys,vers=3,nolock,proto=tcp,comment=cloudconfig       0       2
  /etc/fstab  shdaisreshisensitivedata.blob.core.windows.net:/shdaisreshisensitivedata/ingress  /data   nfs     ro,_netdev,sec=sys,vers=3,nolock,proto=tcp,comment=cloudconfig  0   2
  /etc/fstab  shdaisreshisensitivedata.blob.core.windows.net:/shdaisreshisensitivedata/egress   /output nfs     rw,_netdev,sec=sys,vers=3,nolock,proto=tcp,comment=cloudconfig  0   2
  /etc/fstab  shmdaimysreshimauserdata.file.core.windows.net:/shmdaimysreshimauserdata/shared   /shared nfs     _netdev,sec=sys,nconnect=4,comment=cloudconfig  0       2
  /etc/fstab  shmdaimysreshimauserdata.file.core.windows.net:/shmdaimysreshimauserdata/home     /home   nfs     _netdev,sec=sys,nconnect=4,comment=cloudconfig  0       2
mount.nfs: trying 10.0.1.20 prog 100003 vers 3 prot TCP port 2048
mount.nfs: trying 10.0.1.20 prog 100005 vers 3 prot TCP port 2048
mount.nfs: timeout set for Fri Jul 19 09:41:30 2024
mount.nfs: trying text-based options 'sec=sys,vers=3,nolock,proto=tcp,addr=10.0.1.20'
mount.nfs: prog 100003, trying vers=3, prot=6
mount.nfs: prog 100005, trying vers=3, prot=6
mount.nfs: trying 10.0.1.29 prog 100003 vers 3 prot TCP port 2048
mount.nfs: trying 10.0.1.29 prog 100005 vers 3 prot TCP port 2048
mount.nfs: timeout set for Fri Jul 19 09:41:30 2024
mount.nfs: trying text-based options 'sec=sys,vers=3,nolock,proto=tcp,addr=10.0.1.29'
mount.nfs: prog 100003, trying vers=3, prot=6
mount.nfs: prog 100005, trying vers=3, prot=6
mount.nfs: trying 10.0.1.29 prog 100003 vers 3 prot TCP port 2048
mount.nfs: trying 10.0.1.29 prog 100005 vers 3 prot TCP port 2048
mount.nfs: timeout set for Fri Jul 19 09:41:30 2024
mount.nfs: trying text-based options 'sec=sys,vers=3,nolock,proto=tcp,addr=10.0.1.29'
mount.nfs: prog 100003, trying vers=3, prot=6
mount.nfs: prog 100005, trying vers=3, prot=6
mount.nfs: timeout set for Fri Jul 19 09:41:30 2024
mount.nfs: trying text-based options 'sec=sys,nconnect=4,vers=4.2,addr=10.0.1.28,clientaddr=10.0.2.4'
mount.nfs: timeout set for Fri Jul 19 09:41:30 2024
mount.nfs: trying text-based options 'sec=sys,nconnect=4,vers=4.2,addr=10.0.1.28,clientaddr=10.0.2.4'
/                        : ignored
/boot/efi                : already mounted
/desired_state           : successfully mounted
/data                    : successfully mounted
/output                  : successfully mounted
/shared                  : successfully mounted
/home                    : successfully mounted
mount.nfs: Connection timed out
mount.nfs: Connection timed out
mount.nfs: Connection timed out
```

Maybe because of the NFS changes introduced here?
Can we try rolling those back and open an issue about a stricter configuration (if it works)?.

JimMadge commented:
~~Or, is it only ingress that is failing?~~

jemrobinson commented:
Sure - is it just the port change we need to revert? That seems fine, esp. if we capture it in another issue. BTW, running `mount -fav` manually *does* mount the drives so I'm not sure why this isn't working in the script.

JimMadge commented:
Yes, it is strange, the output it saying timeout. I also saw that on the SRE you deployed when running Ansible, occasional timeout messages.

I don't remember seeing that before, so possibly related to networking changes.

JimMadge committed:
Restore NFS NSG ports

JimMadge commented:
cloudinit still seems to be stalling. Doesn't seem to get past the `mountpoint` loops.

JimMadge committed:
Correct mount check

JimMadge commented:
Found a bug 🤞

JimMadge commented:
Ok, it moves on to running desired state now.

JimMadge commented:
```
[ 3485.382261] desired_state.sh[4934]: PLAY RECAP *********************************************************************
[ 3485.382793] desired_state.sh[4934]: localhost                  : ok=16   changed=11   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
[  OK  ] Finished Run desired state script.
Jul 19 15:02:11 shm-daimyo-sre-oda-vm-workspace-01 systemd[1]: desired-state.service: Deactivated successfully.
Jul 19 15:02:11 shm-daimyo-sre-oda-vm-workspace-01 systemd[1]: Finished Run desired state script.
Jul 19 15:02:11 shm-daimyo-sre-oda-vm-workspace-01 systemd[1]: desired-state.service: Consumed 7min 45.817s CPU time.
[  OK  ] Finished Execute cloud user/final scripts.
[  OK  ] Reached target Cloud-init target.
```

JimMadge commented:
```command
root@shm-daimyo-sre-oda-vm-workspace-01:/# systemctl status desired-state.timer
● desired-state.timer - Run desired state script daily
     Loaded: loaded (/etc/systemd/system/desired-state.timer; enabled; vendor p>
     Active: active (waiting) since Fri 2024-07-19 14:45:36 UTC; 20min ago
    Trigger: Sat 2024-07-20 00:00:00 UTC; 8h left
   Triggers: ● desired-state.service
```

jemrobinson commented on data_safe_haven/resources/workspace/workspace.cloud_init.mustache.yaml:
```diff
@@ -126,7 +126,7 @@ runcmd:
   - mount -fav
   - while (! mountpoint -q /data); do sleep 5; mount /data; done
-  - while (! mountpoint -q /desired_state); do sleep 5; mount /data; done
+  - while (! mountpoint -q /desired_state); do sleep 5; mount /desired_state; done
```
:)

JimMadge replied:
🤦

jemrobinson committed:
Merge branch 'develop' into workspace_software

jemrobinson commented:
1. Cannot log in via GUI with user credentials (grey error box that usually indicates incorrect username/password)
<img width="416" alt="Screenshot 2024-07-21 at 13 19 29" src="https://github.com/user-attachments/assets/483e4e91-a89e-469d-a8bd-5cd94e875e7a">

2. Cannot log in via serial console with admin credentials (not recognised)
<img width="484" alt="Screenshot 2024-07-21 at 13 19 57" src="https://github.com/user-attachments/assets/c51ccd81-bc3b-46a7-92dc-a46520cb40aa">

jemrobinson commented:
Also blob creation in the desired state account is failing on a fresh deployment:

```
  azure-native:storage:Blob (sre_data_blob_desired_state_blob_run_all_tests.bats):
    error: retrieving blob properties "files/home/dshadmin/tests/run_all_tests.bats" (container "desiredstate" / account "shgresrebotdesiredstatec"): blobs.Client#GetProperties: Failure responding to request: StatusCode=403 -- Original Error:
autorest/azure: error response cannot be parsed: {"" '\x00' '\x00'} error: EOF
```

JimMadge commented:
I did see that serial console problem. I did a fresh deployment and it didn't occur again. I assumed it was some problem unrelated to the changes here.

Haven't seen the azure-native Blob error. Looks like an authentication problem.

My output above was from a fresh sre deployment at 226bebd.

jemrobinson commented:
I think the blob issue is actually a Storage Account issue, where sometimes the network permissions on the storage account ("IP address A.B.C.D can access this account") don't translate into *actually* being able to access the account from that IP address. Worth moving this into its own issue as I don't think it's related to this PR.

JimMadge commented:
New deployment at [46af571](https://github.com/alan-turing-institute/data-safe-haven/pull/1892/commits/46af57159d2f0405e3c91bfcac99682ee05e2c79), Pulumi ran without error, Ansible is now running on the workspace VM.

JimMadge commented:
Replicated serial console and guacamole login problems.

#2028 is possibly causing guacamole login problems (redeployed Entra ID App doesn't have the correct token to authenticate with LDAP).

```command
# ldapsearch -H "ldap://identity.oda.daimyo.develop.turingsafehaven.ac.uk:1389"
ldap_sasl_interactive_bind: Protocol error (2)
        additional info: LDAP search request failed. Failed to fetch bearer token from OAuth endpoint.
(unauthorized_client) AADSTS700016: Application with identifier 'd13c364d-992e-4ff2-8911-40c6646c82fd' was not found in the directory 'Turing Data Safe Haven (Green)'. This can happen if the application has not been installed by the administrator of the tenant or consented to by any user in the tenant. You may have sent your authentication request to the wrong tenant. Trace ID: 02613c3e-331e-4e2c-b034-1b317be65200 Correlation ID: 302e5952-a067-4456-95d8-4500577d0602 Timestamp: 2024-07-22 10:49:52Z

JimMadge commented:
For serial console issue, it may be a sporadic problem.

I will try,

- Adding workspaces
- Running deployment without cloudinit and Ansible

I would be surprised if it were cloudinit or Ansible (although I can't think what it _is_) as they shouldn't modify that user and using the portal password reset (either to change `dshadmin`'s password, or create a new user) doesn't work.

JimMadge commented:
Some progress,

I can log in to the serial console using the `dshadmin` account _before_ cloudinit has finished running. That sessions works as long as it is active. If you log out after cloudinit has finished, wait until it has finished before logging in, restart the VM, you get the login incorrect error.

That would suggest there is something in cloud init/ansible that breaks the account.
(Still strange that resetting doesn't appear to work).

JimMadge commented:
In a VM where I had a session,

- `# su dshadmin` works (mimic user without password)
- `login dshadmin` does not work, login incorrect error
- `# passwd dshadmin` lets you change the password, login still fails.

JimMadge commented:
Suspect it may be a PAM configuration issue.

JimMadge committed:
Correct PAM mkhomedir configuration

JimMadge commented:
Before and after correcting `/etc/pam.d/common-session`,

```
Jul 22 14:51:57 shm-daimyo-sre-oda-vm-workspace-01 login[100118]: PAM (other) illegal module type: pam_mkhomedir.so
Jul 22 14:51:57 shm-daimyo-sre-oda-vm-workspace-01 login[100118]: PAM pam_parse: expecting return value; [...skel=/etc/skel]
Jul 22 14:52:01 shm-daimyo-sre-oda-vm-workspace-01 login[100118]: FAILED LOGIN (1) on '/dev/pts/0' FOR 'dshadmin', Permission denied
Jul 22 14:52:09 shm-daimyo-sre-oda-vm-workspace-01 login[100118]: FAILED LOGIN (2) on '/dev/pts/0' FOR 'dshadmin', Permission denied
Jul 22 14:56:49 shm-daimyo-sre-oda-vm-workspace-01 login[101098]: pam_unix(login:session): session opened for user dshadmin(uid=1000) by dshadmin(uid=0)
Jul 22 14:56:54 shm-daimyo-sre-oda-vm-workspace-01 login[101098]: pam_unix(login:session): session closed for user dshadmin
Jul 22 15:02:17 shm-daimyo-sre-oda-vm-workspace-01 login[102510]: pam_unix(login:session): session opened for user dshadmin(uid=1000) by dshadmin(uid=0)
Jul 22 15:02:22 shm-daimyo-sre-oda-vm-workspace-01 login[102510]: pam_unix(login:session): session closed for user dshadmin
```

jemrobinson commented:
For reference, here's `/etc/pam.d/common-session` from an SRE deployed from `develop`

```
#
# /etc/pam.d/common-session - session-related modules common to all services
#
# This file is included from other service-specific PAM config files,
# and should contain a list of modules that define tasks to be performed
# at the start and end of interactive sessions.
#
# As of pam 1.0.1-6, this file is managed by pam-auth-update by default.
# To take advantage of this, it is recommended that you configure any
# local modules either before or after the default block, and use
# pam-auth-update to manage selection of other modules.  See
# pam-auth-update(8) for details.

# here are the per-package modules (the "Primary" block)
session [default=1]                     pam_permit.so
# here's the fallback if no module succeeds
session requisite                       pam_deny.so
# prime the stack with a positive return value if there isn't one already;
# this avoids us returning an error just because nothing sets a success code
# since the modules above will each just jump around
session required                        pam_permit.so
# The pam_umask module will set the umask according to the system default in
# /etc/login.defs and user settings, solving the problem of different
# umask settings with different shells, display managers, remote sessions etc.
# See "man pam_umask".
session optional                        pam_umask.so
# and here are more per-package modules (the "Additional" block)
session required                        pam_unix.so 
session [success=ok default=ignore]     pam_ldap.so minimum_uid=1000
session optional                        pam_systemd.so 
session optional                        pam_mkhomedir.so skel=/etc/skel umask=0022
# end of pam-auth-update config
```

Can you compare to the one deployed from this branch?

JimMadge commented:
@jemrobinson The same (now) 👍
Before there was a line starting with `pam_mkhomedir.so`.

JimMadge committed:
Use community pamd module instead of lineinfile

JimMadge committed:
Use pamd module to handle password change

jemrobinson commented:
I can log into the serial console with these changes and #2034 fixes the Entra application issue. Looking into a problem with the apricot identity server.

jemrobinson commented:
@JimMadge : Interactive login is now working (with the fix from #2034 applied) but I'm getting a prompt that needs sudo approval. Can we get ansible to do this or alternatively, remove it if it's not needed?

<img width="683" alt="Screenshot 2024-07-23 at 08 29 55" src="https://github.com/user-attachments/assets/020f0d6f-9f20-4489-b2d9-bd19ba1372b6">

There's a potential solution here: https://www.reddit.com/r/Ubuntu/comments/15stmwn/how_do_i_suppress_authentication_is_required_to/ but I haven't tested it.

Captured in #2035

JimMadge commented:
@jemrobinson Excellent, thanks for testing.

I'm sure we had that error before and fixed in in v3-4. We should have a look in the old cloud-init, issues/PRs.
I assume you can close the prompt and ignore it safely?

Can we,

1. Merge this and #2034
2. Open an issue for the prompt?

jemrobinson commented:
<img width="110" alt="Screenshot 2024-07-23 at 08 50 47" src="https://github.com/user-attachments/assets/6984f22b-6747-41ca-a791-20dd2d8e8fb3">

looks like we don't have `python-is-python3` installed.

JimMadge committed:
Move python-is-python3 to all releases

JimMadge commented:
I've moved `python-is-python3` to the common list in dfda85d

jemrobinson reviewed (approved):
Happy with this. There are some remaining problems that have been pulled out into their own issues.
